mbtools/models.py

```diff
@@ -58,28 +79,49 @@ def __init__(self, activity_type, activity_path, mbz_path):
         self.activity_filename = \
             str(self.activity_path / f"{self.activity_type}.xml")
         self.etree = etree.parse(self.activity_filename)
+        self.location = ""
 
     def html_elements(self):
         elements = []
         if self.activity_type == "lesson":
-            xpath_query = "//pages/page/contents"
-            elements = self.etree.xpath(xpath_query)
-            xpath_query = "//pages/page/answers/answer_text"
-            elements.extend(self.etree.xpath(xpath_query))
-            return [MoodleHtmlElement(el) for el in elements]
+            lesson_name = self.etree.xpath("//name")[0].text
+            page_query = "//pages/page"
+            pages = self.etree.xpath(page_query)
+            for page in pages:
+                page_id = page.attrib["id"]
+                page_title = page.xpath("title")[0].text
+                contents = page.xpath("contents")
+                location = \
```

Comment:
I totally agree that this method is getting chaotic. So if I understand correctly, the approach you mention means that both MoodleQuestions and MoodleActivities have their own location_string() method that parses the etree itself (I like the latter idea of using the activity type) and each uses the provided etree to generate a location string of the relevant content_type or question. 

That makes sense to me. I think I was thinking previously one pass of a parser was enough but the xpath function is rather efficient I realize.

Reply:
Yeah, the general idea is to push the `location_string()` (or whatever method name) down to the edges as much as possible. Since you can crawl up the tree, I think this would be fairly straight forward, but there may be some gotchas in practice :man_shrugging: .